### ISSUE TITLE
[ERROR] has been blocked by CORS policy

### ISSUE THREAD

**Issue by rojerv** [geckosio/geckos.io#71]
Hello! I'm trying start project on real server (locally working well), my api on api.domain.com  and frontend on domain.com. When i launch and start in browser on https://domain.com i watch this error:

`Access to fetch at 'https://api.domain.com/.wrtc/v1/connections' from origin 'https://domain.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.`

For template i take [this](https://github.com/geckosio/phaser3-multiplayer-game-example) example

My options: 

```js
this.io = geckos({
      iceServers: process.env.NODE_ENV === 'production' ? iceServers : [],
      // cors: { origin: 'https://domain.com' }, // tested too
      cors: {
        origin: (req) => {
          return 'https://domain.com'
        },
      },
    })
```

I hope on help)

**Comment by yandeu** [comment]
I assume if you return `'*'` it works?

api.domain.com and domain.com is not the same. You should return `'https://api.domain.com'` to be able to fetch from `https://api.domain.com/.wrtc/v1/connections`

You can use `req` (short for request) to dynamically return `'https://api.domain.com'` or `'https://domain.com'` depending on the request.

**Comment by rojerv** [comment]
@yandeu  No, same result. I'm started from '*', but same error. When i watch on Network tab i see interesting headers:

OPTIONS `https://api.neonworld.online/.wrtc/v1/connections` return 200
```
Request URL: https://api.neonworld.online/.wrtc/v1/connections
Request Method: OPTIONS
Status Code: 200 
Remote Address: 62.109.3.61:443
Referrer Policy: no-referrer-when-downgrade
```

```
access-control-allow-headers: content-type
access-control-allow-methods: OPTIONS, GET, POST
access-control-allow-origin: *
access-control-request-headers: X-Requested-With, accept, content-type
access-control-request-method: *
content-length: 0
date: Sun, 20 Dec 2020 13:31:20 GMT
```

but second POST `https://api.neonworld.online/.wrtc/v1/connections` return ERROR
and don`t have cors headers. This full list response headers:

```
content-length: 11
content-type: text/plain; charset=utf-8
date: Sun, 20 Dec 2020 13:26:06 GMT
```

**Comment by yandeu** [comment]
> return ERROR

What error? What is the status code?

**Comment by rojerv** [comment]
![Снимок экрана от 2020-12-21 00-09-38](https://user-images.githubusercontent.com/5388627/102715407-f0a08c00-4320-11eb-9352-3461e32f5534.png)


![Снимок экрана от 2020-12-21 00-06-08](https://user-images.githubusercontent.com/5388627/102715346-8be53180-4320-11eb-9734-007342d4521d.png)

![Снимок экрана от 2020-12-21 00-08-32](https://user-images.githubusercontent.com/5388627/102715399-e8485100-4320-11eb-8021-7819ae5c87df.png)

**Comment by yandeu** [comment]
I have tested the URL "https://api.neonworld.online/.wrtc/v1/connections" with a POST request with postman and am getting the status code **502: Bad Gateway**. Are you sure that you forward all post request to the geckos server?

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/502

**Comment by rojerv** [comment]
Yes, current backend is copy paste of [this](https://github.com/geckosio/phaser3-multiplayer-game-example) template. 

samples:

```js
// server.js

const app = express()
const server = http.createServer(app)

const game = new PhaserGame(server)
const port = 4000

// ...

server.listen(port, () => {
  console.log('Express is listening on http://localhost:' + port)
})

```

```js
// gameScene.js

  init() {
    this.io = geckos({
      iceServers: process.env.NODE_ENV === 'production' ? iceServers : [],
      cors: {
        origin: (req) => {
          return '*'
        },
      },
    })
    this.io.addServer(this.game.server)
  }
```

**Comment by yandeu** [comment]
Where are you hosting your app? Most node.js hosting providers (like heroku) do not work, since they don't allow to forward UDP traffic.

I guess the route **.wrtc/v1/connections** fails since it is not possible to make a UDP connection. And this is why the server returns **502: Bad Gateway**.

**Comment by rojerv** [comment]
I'm hosting app on VDS, using traefik 2  for proxy. Trying to set up,  i added udp, but not working yet. I'm still trying to figure it out ...

**Comment by wladikpasika** [comment]
I think, it connected with ports, I have similar an  issue with docker and proxy nginx with https, the  same endpoint falls. As I understand, there shold be available  ports from 0 to 65535(they choosing randomly), and in my case, docker don't allow those ports.

**Comment by yandeu** [comment]
I think it is enough to open 2048-65535/upd (not sure).
Can't this be done with docker like so `docker run -p 2048-65535:2048-65535/udp`?
I have never tested geckos in docker.

The signaling is done on port 9208/tcp. Its traffic can go through proxies.
But the actual WebRTC Channel traffic can't go through proxies. The client connects directly to the server. So the server, where the app is hosted, needs a public IP and open UDP ports.

**Comment by wladikpasika** [comment]
@yandeu I have tried to open ports like this ('2048-65535:2048-65535') but I got lags while docker was building. I opened ports in the range 5000-6000, if more than 1000 - docker just not respond.

**Comment by yandeu** [comment]
It looks like docker takes a long time to open multiple ports.

Maybe you could only open a certain port range. If you know, you will never have more than 200 players, you could try to only open 200 ports.

For example:
- Open port 3000 for your REST API
- Open port 9208 for geckos.io signaling
- Open port 50000-50200/udp for WebRTC

```console
docker run -d -p 3000:3000 -p 9208:9208 -p 50000-50200:50000-50200/udp my-game
```
And on the server, you set the `portRange`:
```js
const io = geckos({
  portRange: {
    min: 50000,
    max: 50200
  }
})
```
_I recently read, the usual port range is 49152–65535._
_I haven't tried this setup yet, but it looks promising!_

**Comment by namlit4** [comment]
I tried to get my example running on mulitple providers with no luck - has anyone succeeded? 
@yandeu do you have a recommendation for hosting?

**Comment by yandeu** [comment]
I personally use AWS EC2.

**Comment by namlit4** [comment]
ok, i was trying to avoid falling to the evil empire until now :-D 
I managed to tie many of your great libs together (headless physics, endable3d-clients and geckos server) and it works beautifully on my machine. but i have been struggling with getting the POC deployed for many days trying three different providers....do you think using socket.io will have a big impact on performance of the setup?

**Comment by meisterpeeps** [comment]
@namlit4 
I've been able to get it working with Digital Ocean.   I have a working game at http://192.241.200.5:1444/
Currently, I'm trying to figure out how I can scale it.  The first thing I tried was using an iframe that the game was embedded in, which I could then spin up different droplets on digital ocean as more people joined, and then just load the iframe that pointed to any particular droplet.  But... I want to use https, and I cant point to an http droplet. And I haven't been able to figure out how to use https with geckos.io.  And if I use https... Ill need domain names for every droplet, which seems problematic.  I have limited knowledge when it comes to setting up the sort of distributed web system that seems like I'd need for it, and geckos.io seems to make it more complicated.    I think my plan would work just fine if I switched my whole system to http, but that doesn't seem satisfying.
